builder/llpcBuilderImplSubgroup.cpp

```diff
@@ -0,0 +1,1185 @@
+
+        // Start the WWM section by setting the inactive invocations.
+        Value* const pSetInactive = CreateSetInactive(pValue, pIdentity);
+
+        // Shift the whole subgroup right by one, using a DPP update operation. This will ensure that the identity
+        // value is in the 0th invocation and all other values are shifted up.
+        Value* const pShiftRight = CreateDPPUpdate(pIdentity, pSetInactive, DPPCtrl::DPP_WF_SR1, 0xf, 0xf, 0);
+
+        Value* pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(2)),
+            CreateGroupOperationArithmetic(groupOperation, pShiftRight,
+                CreateDPPUpdate(pIdentity, pShiftRight, DPPCtrl::DPP_ROW_SR1, 0xf, 0xf, 0)), pShiftRight);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(4)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDPPUpdate(pIdentity, pShiftRight, DPPCtrl::DPP_ROW_SR2, 0xf, 0xf, 0)), pResult);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(4)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDPPUpdate(pIdentity, pShiftRight, DPPCtrl::DPP_ROW_SR3, 0xf, 0xf, 0)), pResult);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(8)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDPPUpdate(pIdentity, pResult, DPPCtrl::DPP_ROW_SR4, 0xf, 0xe, 0)), pResult);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(16)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDPPUpdate(pIdentity, pResult, DPPCtrl::DPP_ROW_SR8, 0xf, 0xc, 0)), pResult);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(32)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDPPUpdate(pIdentity, pResult, DPPCtrl::DPP_ROW_BCAST15, 0xa, 0xf, 0)), pResult);
+
+        pResult = CreateSelect(CreateICmpEQ(pClusterSize, getInt32(64)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDPPUpdate(pIdentity, pResult, DPPCtrl::DPP_ROW_BCAST31, 0xc, 0xf, 0)), pResult);
+
+        // Finish the WWM section by calling the intrinsic.
+        return CreateWWM(pResult);
+    }
+    else
+    {
+        Value* const pThreadMask = CreateThreadMask();
+
+        Value* const pIdentity = CreateGroupOperationIdentity(groupOperation, pValue->getType());
+
+        // Start the WWM section by setting the inactive invocations.
+        Value* const pSetInactive = CreateSetInactive(pValue, pIdentity);
+
+        // Start the WWM section by setting the inactive lanes.
+        Value* pResult = CreateSetInactive(pValue, CreateGroupOperationIdentity(groupOperation, pValue->getType()));
+
+        Value* pMaskedSwizzle = CreateMaskedOperation(pThreadMask, 0xaaaaaaaaaaaaaaaa,
```

Comment:
same as above, please add more comments for the constant

Reply:
I've added loads more comments to explain what is going on!